clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+                    " hostname correctly. You may want to try to adding" +
+                    " '-Dsun.net.spi.nameservice.provider.1=dns,sun' to your client's JVMFLAGS environment." +
+                    " Users must configure FQDN of kafka brokers when authenticating using SASL.";
+            }
+            error += " Kafka Client will go to AUTH_FAILED state.";
+            LOG.error(error);
+            throw new SaslException(error);
+        }
+    }
+
+    private boolean flushNetOutBuffer() throws IOException {
+        if (!netOutBuffer.completed()) {
+            netOutBuffer.writeTo(transportLayer);
+        }
+        return netOutBuffer.completed();
+    }
+
+    public static class ClientCallbackHandler implements CallbackHandler {
+        private String password = null;
+
+        public ClientCallbackHandler(String password) {
+            this.password = password;
+        }
+
+        public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+            for (Callback callback : callbacks) {
+                if (callback instanceof NameCallback) {
+                    NameCallback nc = (NameCallback) callback;
+                    nc.setName(nc.getDefaultName());
+                } else if (callback instanceof PasswordCallback) {
+                    PasswordCallback pc = (PasswordCallback) callback;
+                    if (password != null) {
+                        pc.setPassword(this.password.toCharArray());
+                    } else {
+                        LOG.warn("Could not login: the client is being asked for a password, but the Kafka" +
+                                 " client code does not currently support obtaining a password from the user." +
+                                 " Make sure -Djava.security.auth.login.config property passed to JVM and " +
+                                 " the client is configured to use a ticket cache (using" +
+                                 " the JAAS configuration setting 'useTicketCache=true)'. Make sure you are using" +
+                                 " FQDN of the Kafka broker you are trying to connect to. ");
+                    }
+                } else if (callback instanceof RealmCallback) {
+                    RealmCallback rc = (RealmCallback) callback;
+                    rc.setText(rc.getDefaultText());
+                } else if (callback instanceof AuthorizeCallback) {
+                    AuthorizeCallback ac = (AuthorizeCallback) callback;
+                    String authid = ac.getAuthenticationID();
+                    String authzid = ac.getAuthorizationID();
+
+                    if (authid.equals(authzid))
+                        ac.setAuthorized(true);
+                    else
+                        ac.setAuthorized(false);
+
+
+                    if (ac.isAuthorized())
+                        ac.setAuthorizedID(authzid);
```

Comment:
Looking at the documentation, this only needs to be called if the value passed to `setAuthorizedID` is different than the value of `getAuthorizationID` which is not the case here. Having said that, Hadoop does the same thing so I'll leave it in case it's needed due to non-compliant implementations (unless others disagree).